### PR TITLE
refactor(xds): fold labels into envoy.lb metadata

### PR DESCRIPTION
**Pull request by lukidzi**
## Motivation

When `KUMA_EXPERIMENTAL_INBOUND_TAGS_DISABLED=true` the Dataplane inbound carries only system tags (`kuma.io/zone`, `kuma.io/protocol`); the workload identity that MeshLoadBalancingStrategy affinity relies on is carried to xDS under a separate `io.kuma.labels` endpoint metadata key. That parallel key forced a second extract path and prefer-tags-fall-back-to-labels branching in the locality-aware load balancer. This change removes the parallel representation and folds the labels into the single well-known `envoy.lb` key instead.

## Implementation information

- Fold Dataplane labels into the endpoint tags at topology build time (`endpointIdentity` in `BuildEdsEndpointMap`) when inbound tags are disabled. Inbound tags win over labels on key conflicts, so system tags remain authoritative.
- Thread the flag through the mesh context builder via a new `WithInboundTagsDisabled` option; the control plane wires it from `experimental.inboundTagsDisabled`. With the flag off the generated config is byte-for-byte unchanged.
- Remove the now-redundant `EndpointMetadataWithLabels`, `ExtractLbLabels`, the `io.kuma.labels` metadata key, the `Endpoint.Labels` field, and the endpoint-side prefer-tags-fall-back-to-labels resolver. Net −72 lines.
- Verified safe against every `envoy.lb` / `transport_socket_match` consumer: mTLS SNI derives from cluster/destination tags (not endpoint metadata), and Envoy `transport_socket_match` and subset matches are subset checks, so the extra label keys are inert.

## Supporting documentation

- Affected golden: `locality_aware_inbound_tags_disabled.endpoints.golden.yaml` now carries the workload labels under `envoy.lb` with affinity subzones intact.
- Only relevant when the experimental `KUMA_EXPERIMENTAL_INBOUND_TAGS_DISABLED` flag is set; default behavior is unchanged.

### CODE REVIEW

File: pkg/xds/topology/outbound.go [hunk 1]
```diff
@@ -379,20 +383,41 @@ func fillDataplaneOutbounds(
...
 		}
 	}
 }
 
+// endpointIdentity returns the tags that make up an endpoint's load-balancing
+// identity. When inbound tags are disabled the Dataplane inbound carries only
+// system tags (e.g. kuma.io/zone, kuma.io/protocol), so the workload/resource
+// labels are merged in to preserve label-based identity (affinity, matching).
+// Inbound tags always win over labels on key conflicts.
+func endpointIdentity(inboundTags map[string]string, dataplane *core_mesh.DataplaneResource, inboundTagsDisabled bool) map[string]string {
```
Comment (lukidzi) [inline pkg/xds/topology/outbound.go:374]: Suggestion: The new endpointIdentity function has edge cases (nil tags, empty labels, key conflicts) that could use dedicated unit tests. Current coverage is through integration tests with inboundTagsDisabled=false, but explicit tests for the flag=true path would catch regressions.

File: pkg/xds/topology/outbound_test.go [hunk 2]
```diff
@@ -215,7 +215,7 @@ var _ = Describe("TrafficRoute", func() {
 			externalServices := &core_mesh.ExternalServiceResourceList{}
```
Comment (lukidzi) [inline pkg/xds/topology/outbound_test.go]: Suggestion: Consider adding a table entry with inboundTagsDisabled=true to verify label merging in BuildEdsEndpointMap.

File: pkg/core/xds/types.go [hunk 3]
```diff
@@ -105,13 +105,11 @@ type Endpoint struct {
...
-	// Labels holds resource/workload labels for this endpoint. Unlike Tags (which
-	// are derived from Dataplane inbound configuration), Labels are sourced from
-	// resource metadata (for example, pod labels in Kubernetes mode) and remain
-	// available even when KUMA_EXPERIMENTAL_INBOUND_TAGS_DISABLED is true.
-	Labels          map[string]string
+	// Tags is the endpoint's load-balancing identity. It is derived from the
+	// Dataplane inbound tags; when KUMA_EXPERIMENTAL_INBOUND_TAGS_DISABLED is set
+	// the workload/resource labels are merged in here at topology build time
+	// (see BuildEdsEndpointMap), so a single well-known key carries the identity.
+	Tags            map[string]string
```
Comment (lukidzi) [inline pkg/core/xds/types.go:112]: We should call it metadata as this is the name of the field in envoy endpoint

**Comment by lukidzi** [comment]
Re: renaming `Endpoint.Tags` to `Endpoint.Metadata` — I started the refactor but the scope is significant (90+ files across source and tests). Given this PR's focused scope (folding labels into envoy.lb), I'd suggest making the rename a follow-up PR to keep this one reviewable. The current field name is well-documented in the comment. Happy to do the rename if you'd prefer it bundled here.

File: pkg/xds/topology/outbound.go [hunk 4]
```diff
@@ -379,20 +383,41 @@ func fillDataplaneOutbounds(
...
 				Weight:   endpointWeight,
 				Locality: GetLocality(localZone, getZone(inboundTags), mesh.LocalityAwareLbEnabled()),
 			})
 		}
 	}
 }
 
+// endpointIdentity returns the tags that make up an endpoint's load-balancing
+// identity. When inbound tags are disabled the Dataplane inbound carries only
+// system tags (e.g. kuma.io/zone, kuma.io/protocol), so the workload/resource
```
Comment (lukidzi) [inline pkg/xds/topology/outbound.go:371]: Suggestion: The function allocates an empty map on line 399-401 only to return it immediately when `inboundTagsDisabled` is false. You could reorder — check `inboundTagsDisabled` first and return early with the cloned tags (even if nil), then handle the merge logic. Saves an allocation in the common case.

File: pkg/xds/topology/ingress_outbound_test.go [hunk 5]
```diff
@@ -59,6 +59,7 @@ var _ = Describe("IngressTrafficRoute", func() {
 					egressAddresses,
 					dataSourceLoader,
```
Comment (lukidzi) [inline pkg/xds/topology/ingress_outbound_test.go:59]: Suggestion: All BuildIngressEndpointMap calls pass `false`. Consider adding a test case with `inboundTagsDisabled=true` to verify labels fold into endpoint tags correctly for ingress scenarios.

File: pkg/plugins/policies/meshloadbalancingstrategy/plugin/v1alpha1/locality_aware.go [hunk 6]
```diff
@@ -47,21 +47,16 @@ func NewEndpoints(
...
 	}
-	// Filter each endpoint's Labels to only keys referenced by AffinityTags,
-	// so the rebuilt xDS metadata doesn't carry unrelated pod labels.
-	if conf != nil {
-		for i := range endpointsList {
-			endpointsList[i].Labels = affinityTagPodLabels(endpointsList[i].Labels, *conf)
-		}
-	}
 	return endpoints.ToLocalityLbEndpoints(endpointsList)
 }
```
Comment (lukidzi) [inline pkg/plugins/policies/meshloadbalancingstrategy/plugin/v1alpha1/locality_aware.go:52]: Suggestion: The comment explains that ExtractLbTags returns the identity regardless of source, but doesn't mention that zone tag extraction on line 37 also relies on labels being merged at topology time. The dependency is implicit — might be worth a brief note.

**Review by lukidzi — COMMENTED** [review]
**Verdict: pass**

CI: unit tests and e2e pending; check and build passed. No failures.

No blocking issues. 3 suggestions inline (minor efficiency, test coverage).

File: pkg/xds/envoy/endpoints/v3/endpoints.go [hunk 7]
```diff
@@ -38,7 +38,7 @@ func ToLocalityLbEndpoints(endpoints []core_xds.Endpoint) []*envoy_endpoint.Loca
 			}
 		}
 		lbEndpoint := &envoy_endpoint.LbEndpoint{
-			Metadata: envoy.EndpointMetadataWithLabels(ep.Tags, ep.Labels),
+			Metadata: envoy.EndpointMetadata(ep.Tags),
```
Comment (Automaat) [inline pkg/xds/envoy/endpoints/v3/endpoints.go:41]: **issue:** 8 e2e goldens still contain `io.kuma.labels`, which this line stops emitting: `zoneproxies/meshratelimit/Psni.*` and `zoneproxies/meshtrace/Potel.*` (the `datadog.*`/`zipkin.*` variants were regenerated, these were missed). `MatchGoldenJSON` will fail. Regenerate them.

_Generated by Sybra harness_

File: pkg/plugins/policies/meshloadbalancingstrategy/plugin/v1alpha1/locality_aware.go [hunk 8]
```diff
@@ -113,7 +109,7 @@ func configureCrossZoneEndpointLocality(crossZonePriorityGroups []CrossZoneLbGro
 
 func configureLocalZoneEndpointLocality(localPriorityGroups []LocalLbGroup, endpoint *core_xds.Endpoint, localZone string) {
 	for _, localRule := range localPriorityGroups {
-		val, ok := resolveEndpointAffinityValue(endpoint.Tags, endpoint.Labels, localRule.Key)
+		val, ok := endpoint.Tags[localRule.Key]
```
Comment (Automaat) [inline pkg/plugins/policies/meshloadbalancingstrategy/plugin/v1alpha1/locality_aware.go:112]: **issue:** `priority.go` still falls back tags->pod-labels to resolve affinity, but this line now reads `endpoint.Tags` only. With the flag off, labels aren't folded, so an AffinityTag key present only in pod labels silently stops matching. Make both sides consistent, or confirm flag-off affinity keys are always inbound tags.

_Generated by Sybra harness_

File: pkg/core/bootstrap/bootstrap.go [hunk 9]
```diff
@@ -517,6 +517,9 @@ func initializeMeshCache(builder *core_runtime.Builder) error {
 	if builder.Config().XdsServer.PolicyMatchingCacheSize > 0 {
 		mcbOpts = append(mcbOpts, xds_context.WithPolicyMatchingHash())
 	}
+	if builder.Config().Experimental.InboundTagsDisabled {
+		mcbOpts = append(mcbOpts, xds_context.WithInboundTagsDisabled(true))
```
Comment (Automaat) [inline pkg/core/bootstrap/bootstrap.go:521]: **issue:** Flag wired into the xDS mesh-cache builder only. The api-server's separate `NewMeshContextBuilder` (server.go:~520) isn't passed it, so with the flag on the inspect API builds endpoints without folded labels, diverging from served xDS. Thread it there too.

_Generated by Sybra harness_

File: pkg/xds/topology/outbound_test.go [hunk 10]
```diff
@@ -262,6 +262,94 @@ var _ = Describe("TrafficRoute", func() {
 				},
 			}))
 		})
+
+		It("should merge dataplane labels into endpoint tags when inboundTagsDisabled=true", func() {
```
Comment (Automaat) [inline pkg/xds/topology/outbound_test.go:263]: **suggestion:** These fold tests use synthetic labels (`app`/`version`/`env`). Real `Meta.Labels` also carries `kuma.io/display-name`, `kuma.io/mesh`, `k8s.kuma.io/namespace`, etc., all folded into `envoy.lb`. Add a case with realistic `kuma.io/*` labels to pin that behavior.

_Generated by Sybra harness_

**Review by Automaat — CHANGES_REQUESTED** [review]
Solid, well-scoped refactor — collapsing the parallel `io.kuma.labels` path into a single `envoy.lb` key is a real simplification, and the flag plumbing mirrors the existing `WithPolicyMatchingHash` pattern. Requesting changes on two items before merge: 8 e2e goldens (`Psni.*`, `Potel.*`) still carry `io.kuma.labels` and will fail `MatchGoldenJSON`, and the endpoint-side affinity match dropped the pod-label fallback while `priority.go` kept it — confirm that can't silently break flag-off label-based affinity. 3 issues, 1 suggestion.

_Generated by Sybra harness_

**Review by Automaat — DISMISSED** [review]

**Review by Automaat — APPROVED** [review]
